chatmaild/src/chatmaild/metadata.py

```diff
@@ -18,21 +21,42 @@ def __init__(self, vmail_dir):
     def get_metadata_dict(self, addr):
         return FileDict(self.vmail_dir / addr / "metadata.json")
 
-    def add_token_to_addr(self, addr, token):
+    @contextmanager
+    def _modify_tokens(self, addr):
         with self.get_metadata_dict(addr).modify() as data:
-            tokens = data.setdefault(self.DEVICETOKEN_KEY, [])
-            if token not in tokens:
-                tokens.append(token)
+            tokens = data.setdefault(self.DEVICETOKEN_KEY, {})
+            now = int(time.time())
+            if isinstance(tokens, list):
+                data[self.DEVICETOKEN_KEY] = tokens = {t: now for t in tokens}
+
+            # Remove tokens that are older than a week
+            # or have invalid timestamp from the future.
+            expired_tokens = [
+                token
+                for token, timestamp in tokens.items()
+                if now > tokens[token] + 3600 * 24 * 7 or tokens[token] > now + 60
```

Comment:
```suggestion
                if tokens[token] < now - 3600 * 24 * 7 or tokens[token] > now + 60
```

I find your option a bit harder to read; let's always compare the token with the modification of `now`.